File: scripts/BinRead.py
#!/usr/bin/env python3
import numpy as np
from os import environ, path
import sys
import struct
from argparse import ArgumentParser

def BinRead(infile, outfile='', verbose=0):  #{{{

    print('reading binary file')
    f = open(infile, 'rb')

    if outfile:
        sys.stdout = open(outfile, 'w')

    while True:
        try:
            # Step 1: Read size of record name
            recordnamesize = struct.unpack('i', f.read(struct.calcsize('i')))[0]
        except struct.error as e:
            print('probable EOF: {}'.format(e))
            break

        print('============================================================================')
        if verbose > 2:
            print('\n recordnamesize = {}'.format(recordnamesize))
        recordname = struct.unpack('{}s'.format(recordnamesize), f.read(recordnamesize))[0].decode('ASCII')
        print('field: {}'.format(recordname))

        # Step 2: Read the data itself
        # First read length of record
        #reclen = struct.unpack('i', f.read(struct.calcsize('i')))[0]
        reclen = struct.unpack('q', f.read(struct.calcsize('q')))[0]
        if verbose > 1:
            print('reclen = {}'.format(reclen))

        # Read data code
        code = struct.unpack('i', f.read(struct.calcsize('i')))[0]
        print('Format = {} (code {})'.format(CodeToFormat(code), code))

        if code == FormatToCode('Boolean'):
            bval = struct.unpack('i', f.read(reclen - struct.calcsize('i')))[0]
            print('value = {}'.format(bval))

        elif code == FormatToCode('Integer'):
            ival = struct.unpack('i', f.read(reclen - struct.calcsize('i')))[0]
            print('value = {}'.format(ival))

        elif code == FormatToCode('Double'):
            dval = struct.unpack('d', f.read(reclen - struct.calcsize('i')))[0]
            print('value = {}'.format(dval))

        elif code == FormatToCode('String'):
            strlen = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            if verbose > 1:
                print('strlen = {}'.format(strlen))
            sval = struct.unpack('{}s'.format(strlen), f.read(strlen))[0]
            print('value = {}'.format(sval))

        elif code == FormatToCode('BooleanMat'):
            # Read matrix type
            mattype = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            print('mattype = {}'.format(mattype))

            # Read matrix
            s = [0, 0]
            s[0] = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            s[1] = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            print('size = [{}x{}]'.format(s[0], s[1]))
            data = np.zeros((s[0], s[1]))
            for i in range(s[0]):
                for j in range(s[1]):
                    data[i][j] = struct.unpack('d', f.read(struct.calcsize('d')))[0]
                    if verbose > 2:
                        print('data[{}, {}] = {}'.format(i, j, data[i][j]))

        elif code == FormatToCode('IntMat'):
            # Read matrix type
            mattype = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            print('mattype = {}'.format(mattype))

            # Read matrix
            s = [0, 0]
            s[0] = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            s[1] = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            print('size = [{}x{}]'.format(s[0], s[1]))
            data = np.zeros((s[0], s[1]))
            for i in range(s[0]):
                for j in range(s[1]):
                    data[i][j] = struct.unpack('d', f.read(struct.calcsize('d')))[0]
                    if verbose > 2:
                        print('data[{}, {}] = {}'.format(i, j, data[i][j]))

        elif code == FormatToCode('DoubleMat'):
            # Read matrix type
            mattype = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            print('mattype = {}'.format(mattype))

            # Read matrix
            s = [0, 0]
            s[0] = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            s[1] = struct.unpack('i', f.read(struct.calcsize('i')))[0]
            print('size = [{}x{}]'.format(s[0], s[1]))
            data = np.zeros((s[0], s[1]))
            for i in range(s[0]):
                for j in range(s[1]):
                    data[i][j] = struct.unpack('d', f.read(struct.calcsize('d')))[0]
                    if verbose > 2:
                        print('data[{}, {}] = {}'.format(i, j, data[i][j]))

        elif code == FormatToCode('MatArray'):
            f.seek(reclen - 4, 1)
            print('skipping {} bytes for code {}'.format(code, reclen - 4))
        elif code == FormatToCode('StringArray'):
            f.seek(reclen - 4, 1)
            print('skipping {} bytes for code {}'.format(code, reclen - 4))

        else:
            raise TypeError('BinRead error message: data type: {} not supported yet! ({})'.format(code, recordname))

    f.close()
#}}}

def FormatToCode(format):  # {{{
    """This routine takes the format string and converts it into an integer, 
    which is passed along with the record in order to identify the nature of 
    the data being sent.
    """

    if format == 'Boolean':
        code = 1
    elif format == 'Integer':
        code = 2
    elif format == 'Double':
        code = 3
    elif format == 'String':
        code = 4
    elif format == 'BooleanMat':
        code = 5
    elif format == 'IntMat':
        code = 6
    elif format == 'DoubleMat':
        code = 7
    elif format == 'MatArray':
        code = 8
    elif format == 'StringArray':
        code = 9
    else:
        raise IOError('FormatToCode error message: data type not supported yet!')

    return code
# }}}

def CodeToFormat(code):  # {{{
    """This routine takes a datatype code and converts it to the corresponding 
    string in order to identify the nature of the data retrieved.
    """

    if code == 1:
        format = 'Boolean'
    elif code == 2:
        format = 'Integer'
    elif code == 3:
        format = 'Double'
    elif code == 4:
        format = 'String'
    elif code == 5:
        format = 'BooleanMat'
    elif code == 6:
        format = 'IntMat'
    elif code == 7:
        format = 'DoubleMat'
    elif code == 8:
        format = 'MatArray'
    elif code == 9:
        format = 'StringArray'
    else:
        raise TypeError('FormatToCode error message: code {} not supported yet!'.format(code))

    return format
# }}}

if __name__ == '__main__':  #{{{
    parser = ArgumentParser(description='BinRead - function to read binary input file.')
    parser.add_argument('-f', '--infile', help='name of binary input file', default='')
    parser.add_argument('-o', '--outfile', help='optional name of text output file', default='')
    parser.add_argument('-v', '--verbose', help='optional level of output', default=0)
    args = parser.parse_args()

    BinRead(args.infile, args.outfile, args.verbose)
#}}}

File: src/m/solve/BinRead.m
function md = BinReadMatlab(infile)
%BINREADMATLAB - Read ISSM input binary 
%
%   struct = BinReadMatlab(infile)

	% Open binary file for reading
	fid = fopen(infile, 'rb');
	if fid == -1
		error('Cannot open input file %s', infile);
	end

	%Prepare output
	md = model();

	while(true)
		% Step 1: read size of record name (int32)
		[recordNameSize, cnt] = fread(fid, 1, 'int32');
		if cnt ~= 1
			% EOF reached
			break;
		end

		% Read the record name 
		rawName = fread(fid, recordNameSize, '*char')';
		recordName = char(rawName);
		fprintf('Reading %s\n', recordName);

		%See if field exists in md
		parts =  strsplit(recordName, '.');
		addtomd = true;
		if numel(parts)<3
			addtomd = false;
		elseif ~isprop(md, parts{2})
			addtomd = false;
		elseif ~isprop(md.(parts{2}), parts{3})
			addtomd = false;
		end


		% Step 2: read length of record (int64)
		reclen = fread(fid, 1, 'int64');

		% Read data code (int32)
		code = fread(fid, 1, 'int32');
		formatStr = CodeToFormat(code);
		%fprintf('Format = %s (code %d)\n', formatStr, code);

		% Dispatch based on the data type
		switch code
			case FormatToCode('Boolean')
				% Boolean stored as 32‑bit integer (0/1)
				data = fread(fid, 1, 'int32');

			case FormatToCode('Integer')
				data = fread(fid, 1, 'int32');

			case FormatToCode('Double')
				data = fread(fid, 1, 'double');

			case FormatToCode('String')
				len  = fread(fid, 1, 'int');
				data = char(fread(fid, len, 'char'))';

			case {FormatToCode('BooleanMat'), FormatToCode('IntMat'), FormatToCode('DoubleMat')}

				% Read matrix type (int32) – not used further here
				mattype = fread(fid, 1, 'int32');

				% Read dimensions (two int32s)
				rows = fread(fid, 1, 'int32');
				cols = fread(fid, 1, 'int32');

				% Choose element precision according to the specific matrix type
				switch code
					case FormatToCode('BooleanMat')
						elemType = 'double';
					case FormatToCode('IntMat')
						elemType = 'int32';
					case FormatToCode('DoubleMat')
						elemType = 'double';
				end

				% Read the matrix data column‑major (Python wrote row‑major;
				% MATLAB stores column‑major, so we transpose after reading)
				data = fread(fid, rows*cols, elemType);
				data = reshape(data, [cols, rows])';


			otherwise
				% For unsupported or array types we simply skip the remaining bytes
				% (same behaviour as the Python script)
				skipBytes = reclen - 4;  % 4 bytes already consumed by the code
				fseek(fid, skipBytes, 'cof');
				fprintf('skipping %d bytes for code %d\n', skipBytes, code);
		end
		if addtomd
			md.(parts{2}).(parts{3}) = data;
		else
			disp([' -> Skipping ', recordName]);
		end

		%Change class depending on what's provided
		if strcmp(recordName, 'md.mesh.elementtype')
			switch(data)
				case 'Tria';  md.mesh = mesh2d();
				case 'Penta'; md.mesh = mesh3dprisms();
				otherwise error(['Element type ', data, ' not supported yet']);
			end
		end
	end
	fclose(fid);
end

function fmt = CodeToFormat(code) % {{{
    switch code
        case 1, fmt = 'Boolean';
        case 2, fmt = 'Integer';
        case 3, fmt = 'Double';
        case 4, fmt = 'String';
        case 5, fmt = 'BooleanMat';
        case 6, fmt = 'IntMat';
        case 7, fmt = 'DoubleMat';
        case 8, fmt = 'MatArray';
        case 9, fmt = 'StringArray';
        otherwise, error('Unsupported code: %d', code);
    end
end % }}}
function code=FormatToCode(format) % {{{
	%This routine takes the format string, and hardcodes it into an integer, which
	%is passed along the record, in order to identify the nature of the dataset being
	%sent.
	if     strcmpi(format,'Boolean'),
		code=1;
	elseif strcmpi(format,'Integer'),
		code=2;
	elseif strcmpi(format,'Double'),
		code=3;
	elseif strcmpi(format,'String'),
		code=4;
	elseif strcmpi(format,'BooleanMat'),
		code=5;
	elseif strcmpi(format,'IntMat'),
		code=6;
	elseif strcmpi(format,'DoubleMat'),
		code=7;
	elseif strcmpi(format,'MatArray'),
		code=8;
	elseif strcmpi(format,'StringArray'),
		code=9;
	elseif strcmpi(format,'CompressedMat'),
		code=10;
	else
		error('FormatToCode error message: data type not supported yet!');
	end
end% }}}
